apps/prairielearn/assets/scripts/lib/scatter.ts

```diff
@@ -0,0 +1,183 @@
+  ydata = ydata.map((y) => Math.max(ymin, Math.min(ymax, y)));
+
+  let xAxis = axisBottom(x).tickValues(resolvedOptions.xgrid);
+  if (resolvedOptions.xTickLabels !== 'auto') {
+    xAxis = xAxis.tickFormat((d, i) => resolvedOptions.xTickLabels[i]);
+  }
+
+  let yAxis = axisLeft(y).tickValues(resolvedOptions.ygrid);
+  if (resolvedOptions.yTickLabels !== 'auto') {
+    yAxis = yAxis.tickFormat((d, i) => resolvedOptions.yTickLabels[i]);
+  }
+
+  const xGrid = axisBottom(x)
+    .tickValues(resolvedOptions.xgrid)
+    .tickSize(-height)
+    .tickFormat(() => '');
+
+  const yGrid = axisLeft(y)
+    .tickValues(resolvedOptions.ygrid)
+    .tickSize(-width)
+    .tickFormat(() => '');
+
+  const svg = select(selector)
+    .append('svg')
+    .attr('width', width + (resolvedOptions.leftMargin ?? 70) + (resolvedOptions.rightMargin ?? 20))
+    .attr(
+      'height',
+      height + (resolvedOptions.topMargin ?? 10) + (resolvedOptions.bottomMargin ?? 55),
+    )
+    .attr('class', 'center-block statsPlot')
+    .append('g')
+    .attr(
+      'transform',
+      'translate(' + resolvedOptions.leftMargin + ',' + resolvedOptions.topMargin + ')',
+    );
+
+  svg
+    .append('g')
+    .attr('class', 'x grid')
+    .attr('transform', 'translate(0,' + height + ')')
+    .call(xGrid);
+
+  svg.append('g').attr('class', 'y grid').call(yGrid);
+
+  svg
+    .append('g')
+    .attr('class', 'x axis')
+    .attr('transform', 'translate(0,' + height + ')')
+    .call(xAxis)
+    .append('text')
+    .attr('class', 'label')
+    .attr('x', width / 2)
+    .attr('y', '3em')
+    .style('text-anchor', 'middle')
+    .text(resolvedOptions.xlabel);
+
+  svg
+    .append('g')
+    .attr('class', 'y axis')
+    .call(yAxis)
+    .append('text')
+    .attr('class', 'label')
+    .attr('transform', 'rotate(-90)')
+    .attr('x', -height / 2)
+    .attr('y', '-3em')
+    .style('text-anchor', 'middle')
+    .text(resolvedOptions.ylabel);
+
+  // zips the data used to create the scatter plot
+  // each data point has value [x, y, label]
+  // if options.labels is not specified, then each data point will have an undefined label, and no labels will appear in the plot
+  let pData = zip(xdata, ydata, resolvedOptions.labels);
+
+  // if the number of labels is not equal to the number of data points, not all the data points will render. In this case, we zip the data points together without labels to preserve data.
+  if (
+    resolvedOptions.labels.length !== xdata.length ||
+    resolvedOptions.labels.length !== ydata.length
+  ) {
+    pData = zip(xdata, ydata);
+  }
+
+  const nodes = svg.selectAll('.point').data(pData).enter().append('g');
+
+  nodes
+    .append('circle')
+    .attr('class', 'point')
+    .attr('cx', function (p: number[]) {
+      return x(p[0]);
+    })
+    .attr('cy', function (p: number[]) {
+      return y(p[1]);
+    })
+    .attr('r', function () {
+      return resolvedOptions.radius;
+    });
```

Comment:
You can simplify this and the following one in two ways:

- Arrow functions
- Implicit types

Example:

```suggestion
  nodes
    .append('circle')
    .attr('class', 'point')
    .attr('cx', (p) => x(p[0]))
    .attr('cy', (p) => y(p[1]))
    .attr('r', () => resolvedOptions.radius);
```

I'll leave the below `nodes.append('text')...` as an exercise for you!

Reply:
Updated here and `nodes.append('text')...` in [4dad954](https://github.com/PrairieLearn/PrairieLearn/pull/9809/commits/4dad954294139e1ccb0018548e102663ab43c2df).